src/__swaps__/screens/Swap/providers/swap-provider.tsx

```diff
@@ -490,63 +501,51 @@ export const SwapProvider = ({ children }: SwapProviderProps) => {
     };
   }, []);
 
-  const confirmButtonIcon = useDerivedValue(() => {
+  const gasSettings = useSharedValue<GasSettings | undefined>(undefined);
+  const estimatedGasLimit = useSharedValue<string | undefined>(undefined);
+  const userHasEnoughFundsForTx = useSharedValue<boolean>(true);
+
+  const confirmButtonProps = useDerivedValue(() => {
     if (isSwapping.value) {
-      return '';
+      return { label: 'Swapping...', disabled: true };
     }
 
     if (configProgress.value === NavigationSteps.SHOW_REVIEW) {
-      return '􀎽';
-    } else if (configProgress.value === NavigationSteps.SHOW_GAS) {
-      return '􀆅';
+      return { icon: '􀎽', label: 'Hold to Swap', disabled: false };
+    }
+
+    if (configProgress.value === NavigationSteps.SHOW_GAS) {
+      return { icon: '􀆅', label: 'Save', disabled: false };
     }
 
-    if (isQuoteStale.value === 1 && sliderPressProgress.value === 0) {
-      return '';
+    if (isFetching.value) {
+      return { label: 'Fetching...', disabled: true };
     }
 
+    const hasSelectedAssets = internalSelectedInputAsset.value && internalSelectedOutputAsset.value;
+    if (!hasSelectedAssets) return { label: 'Select Token', disabled: true };
+
+    if (!quote.value || 'error' in quote.value) return { label: 'Error', disabled: true };
+
     const isInputZero = equalWorklet(SwapInputController.inputValues.value.inputAmount, 0);
     const isOutputZero = equalWorklet(SwapInputController.inputValues.value.outputAmount, 0);
 
-    if (
-      (isInputZero && isOutputZero) ||
-      isFetching.value ||
-      (SwapInputController.inputMethod.value === 'slider' && SwapInputController.percentageToSwap.value === 0)
-    ) {
-      return '';
-    } else {
-      return '􀕹';
+    if (SwapInputController.percentageToSwap.value === 0 || isInputZero || isOutputZero) {
+      return { label: 'Enter Amount', disabled: true };
     }
-  });
 
-  const confirmButtonLabel = useDerivedValue(() => {
-    if (isSwapping.value) {
-      return swapping;
-    }
+    if (!estimatedGasLimit.value) return { label: 'Estimating...', disabled: true };
 
-    if (configProgress.value === NavigationSteps.SHOW_REVIEW) {
-      return tapToSwap;
-    } else if (configProgress.value === NavigationSteps.SHOW_GAS) {
-      return save;
+    if (!gasSettings.value) {
+      // this could happen if metereology is down, or some other edge cases that are not properly handled yet
+      return { label: 'Error', disabled: true };
     }
 
-    if (isFetching.value || (isQuoteStale.value === 1 && SwapInputController.inputMethod.value !== 'slider')) {
-      return fetchingPrices;
+    if (!userHasEnoughFundsForTx.value) {
```

Comment:
I'm noticing an `Insufficient Funds` flash in the video; generally we want to avoid states of flashing / flickering